### DISCUSSION TITLE
16-bit palette

### DISCUSSION THREAD

**Discussion by Manwe-SandS** [ImageMagick/ImageMagick#2559] [Help]
To code a demo for retro-computer, I need to convert true-color image to _256 out of 65536_ colors. Palette format of that retro computer is 5-6-5 bits (for red, green, blue).
I found this ImageMagick's option:

`convert input.png -dither FloydSteinberg -colors 256 -depth 5 output.bmp`

I expect to see zeros in 3 lowest bits of every color component, since I set 5-bit depth (Ok, it's not 5-6-5, but 5-5-5 is better then nothing). In output.bmp I've got 32 shades in each colour channel, but palette contains full 8 bit values (every bit may be set to 1).

Is there any way to keep palette's values in 5 or 6 bits? May be some specific color space or so?

**Comment by snibgo** [comment]
Reducing to RGB555 (ignoring the next problem, of reducing colours and creating a palette) is easy. Assuming "magick" is Q16, Windows format:
```
magick ^
  in.png ^
  -alpha off ^
  -evaluate Divide 257 ^
  -evaluate And 248 ^
  -evaluate Multiply 257 ^
  -depth 8 ^
  out.png
```
Note that decimal 248 is the 8-bit binary number 11111000. By using a different AND for the channels, we can get RGB565:
```
magick ^
  toes.png ^
  -alpha off ^
  -evaluate Divide 257 ^
  -channel RB -evaluate And 248 ^
  -channel G -evaluate And 252 ^
  +channel ^
  -evaluate Multiply 257 ^
  -depth 8 ^
  x.png
```
The result could have 65536 different colours. Reducing this with "-colors" won't work, because quantization can introduce new colours.

https://www.imagemagick.org/script/command-line-options.php#kmeans suggests we can:
```
magick ^
  x.png ^
  -define kmeans:seed-colors=black ^
  -kmeans 255 ^
  x1.png
```
... to avoid introducing new colours. However, this does introduce new colours.

**Comment by snibgo** [comment]
On the other hand, perhaps this directly does what you want:
```
magick in.png -define bmp:subtype=RGB565 out.bmp
```

**Comment by Manwe-SandS** [comment]
Wow, "And" and "Divide" are the powerful options! Thank you.
